File: app/src/main/java/com/github/benji377/timety/utils/InsightsGenerator.kt
package com.github.benji377.timety.utils

import com.github.benji377.timety.data.FocusSession
import java.util.Calendar

/**
 * Generates insights based on focus session data.
 */
object InsightsGenerator {

    fun generateInsights(sessions: List<FocusSession>): List<String> {
        if (sessions.isEmpty()) {
            return listOf("Start focusing to unlock insights!")
        }

        val insights = mutableListOf<String>()

        // Insight 1: Total focus time
        val totalMinutes = sessions.sumOf { it.duration / 60000L }
        if (totalMinutes > 0) {
            insights.add("You've spent $totalMinutes minutes focusing 🎯")
        }

        // Insight 2: Best time of day
        val timeOfDayBreakdown = sessions.groupBy { session ->
            val cal = Calendar.getInstance().apply { timeInMillis = session.startTime }
            when (cal.get(Calendar.HOUR_OF_DAY)) {
                in 5..11 -> "morning"
                in 12..17 -> "afternoon"
                in 18..21 -> "evening"
                else -> "night"
            }
        }

        val bestTimeOfDay =
            timeOfDayBreakdown.maxByOrNull { it.value.sumOf { s -> s.duration } }?.key
        if (bestTimeOfDay != null) {
            val timeLabel = when (bestTimeOfDay) {
                "morning" -> "Morning"
                "afternoon" -> "Afternoon"
                "evening" -> "Evening"
                else -> "Night"
            }
            insights.add("You're most focused in the $timeLabel ⏰")
        }

        // Insight 3: Streak milestone
        val recentSessions = sessions.sortedByDescending { it.startTime }.take(7)
        if (recentSessions.size >= 5) {
            insights.add("You're on a focus spree! Keep it up 🔥")
        }

        // Insight 4: Average session length
        val avgSessionMinutes = (sessions.sumOf { it.duration } / sessions.size) / 60000L
        if (avgSessionMinutes > 0) {
            val sessionLabel =
                if (avgSessionMinutes < 30) "short" else if (avgSessionMinutes < 90) "medium" else "long"
            insights.add("Your average focus session is $sessionLabel (~$avgSessionMinutes min)")
        }

        // Insight 5: Rating analysis
        val ratings = sessions.groupBy { it.rating }
        val highRatingPercentage = ratings.filterKeys { it?.name == "GREAT" }
            .values.sumOf { it.size } * 100 / sessions.size
        if (highRatingPercentage >= 70) {
            insights.add("Great focus quality! You're consistently in the zone 💪")
        } else if (highRatingPercentage < 30) {
            insights.add("Try shorter sessions for better focus quality 🎯")
        }

        return insights.take(3) // Return top 3 insights
    }
}

File: lib/utils/insights_generator.dart
import '../data/focus_session.dart';

class InsightsGenerator {
  static List<String> generateInsights(List<FocusSession> sessions) {
    if (sessions.isEmpty) {
      return ["Start focusing to unlock insights!"];
    }

    final insights = <String>[];

    // Insight 1: Total focus time
    final totalMinutes = sessions.fold(0, (sum, s) => sum + (s.duration ~/ 60000));
    if (totalMinutes > 0) {
      insights.add("You've spent $totalMinutes minutes focusing 🎯");
    }

    // Insight 2: Best time of day
    final timeOfDayBreakdown = <String, int>{};
    for (var session in sessions) {
      final hour = DateTime.fromMillisecondsSinceEpoch(session.startTime).hour;
      String timeOfDay;
      if (hour >= 5 && hour <= 11) {
        timeOfDay = "morning";
      } else if (hour >= 12 && hour <= 17) {
        timeOfDay = "afternoon";
      } else if (hour >= 18 && hour <= 21) {
        timeOfDay = "evening";
      } else {
        timeOfDay = "night";
      }
      timeOfDayBreakdown[timeOfDay] = (timeOfDayBreakdown[timeOfDay] ?? 0) + session.duration;
    }

    if (timeOfDayBreakdown.isNotEmpty) {
      final bestTimeOfDay = timeOfDayBreakdown.entries.reduce((a, b) => a.value > b.value ? a : b).key;
      final timeLabel = bestTimeOfDay[0].toUpperCase() + bestTimeOfDay.substring(1);
      insights.add("You're most focused in the $timeLabel ⏰");
    }

    // Insight 3: Streak milestone
    if (sessions.length >= 5) {
      insights.add("You're on a focus spree! Keep it up 🔥");
    }

    // Insight 4: Average session length
    final avgSessionMinutes = totalMinutes ~/ sessions.length;
    if (avgSessionMinutes > 0) {
      final sessionLabel = avgSessionMinutes < 30 ? "short" : (avgSessionMinutes < 90 ? "medium" : "long");
      insights.add("Your average focus session is $sessionLabel (~$avgSessionMinutes min)");
    }

    return insights.take(3).toList();
  }
}
